terraform-tests/mssql_test.go

```diff
@@ -63,4 +64,200 @@ var _ = Describe("mssql", Label("mssql-terraform"), Ordered, func() {
 			}))
 		})
 	})
+
+	Context("properties that pass-through without validation", func() {
+		When("mssql_version is passed", func() {
+			It("it passes-through without validation", func() {
+				// It would be possible to validate the engine version before apply using
+				// https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/rds_engine_version
+				// It has some very interesting properties and use cases, such as:
+				//  - valid_upgrade_targets - Set of engine versions that this database engine version can be upgraded to.
+				//  - supports_read_replica - Indicates whether the database engine version supports read replicas.
+				//  - supports_log_exports_to_cloudwatch - Indicates whether the engine version supports exporting the log types specified by
+				plan = ShowPlan(terraformProvisionDir, buildVars(defaultVars, map[string]any{"mssql_version": "ANY-VALUE-AT-ALL"}))
+				Expect(AfterValuesForType(plan, "aws_db_instance")).To(MatchKeys(IgnoreExtras, Keys{"engine_version": Equal("ANY-VALUE-AT-ALL")}))
+			})
+		})
+		When("instance_class is passed", func() {
+			It("it passes-through without validation", func() {
+				plan = ShowPlan(terraformProvisionDir, buildVars(defaultVars, map[string]any{"instance_class": "ANY-VALUE-AT-ALL"}))
+				Expect(AfterValuesForType(plan, "aws_db_instance")).To(MatchKeys(IgnoreExtras, Keys{"instance_class": Equal("ANY-VALUE-AT-ALL")}))
+			})
+		})
+	})
+	Context("instance_name", func() {
+		When("invalid instance_name is passed", func() {
+			It("fails and returns a descriptive message", func() {
+				session, _ := FailPlan(terraformProvisionDir, buildVars(defaultVars, map[string]any{"instance_name": "THIS-ENGINE-DOESNT-EXIST"}))
+
+				Expect(session.ExitCode()).NotTo(Equal(0))
+				Expect(session.Err).To(gbytes.Say("only lowercase alphanumeric characters, hyphens, underscores, periods, and spaces allowed in \"name\""))
+
+			})
+		})
+		When("instance_name is passed", func() {
+			It("it is used to populate several fields", func() {
+				plan = ShowPlan(terraformProvisionDir, buildVars(defaultVars, map[string]any{"instance_name": "some-better-value"}))
+				Expect(AfterValuesForType(plan, "aws_db_instance")).To(MatchKeys(IgnoreExtras, Keys{"identifier": Equal("some-better-value")}))
+				Expect(AfterValuesForType(plan, "aws_security_group")).To(MatchKeys(IgnoreExtras, Keys{"name": Equal("some-better-value-sg")}))
+				Expect(AfterValuesForType(plan, "aws_db_subnet_group")).To(MatchKeys(IgnoreExtras, Keys{"name": Equal("some-better-value-p-sn")}))
+			})
+		})
+	})
+
+	Context("engine", func() {
+		When("no engine passed", func() {
+			It("should use the default value", func() {
+				plan = ShowPlan(terraformProvisionDir, buildVars(defaultVars, nil))
+				Expect(AfterValuesForType(plan, "aws_db_instance")).To(MatchKeys(IgnoreExtras, Keys{"engine": Equal("sqlserver-ee")}))
+			})
+		})
+
+		When("valid engine passed", func() {
```

Comment:
Most of these tests are duplicated in integration and terraform tests due to having the validation in both places. Reasonings for leaving validation in yaml:

- documentation purposes. 
- faster feedback: variable resolution happens before spinning out the go routine that ends up calling terraform, means validation written in the yaml file and executed by the broker will provide feedback syncrhonously to the user.

We can continue thinking on shifting where we validate. I would like that to be done weighting pros and cons and allowing for us to raise concerns as a team.

Reply:
I'll wait for that proposal, it is a very interesting discussion to have.